### PR TITLE
Transition to Girder plugin

### PR DESCRIPTION
**Pull request by danlamanna**
- [X] Get Flow installable as a Girder plugin
- [x] Potentially update provisioning scripts for installing Flow
- [x] Get travis working with and without R
- [x] Move to Girder's testing infrastructure
   -  [X] Remove Tangelo from serving tests

There are still issues remaining with styling and fonts that need to be fixed, as well as exhaustive user testing of the system.

### CODE REVIEW

**Comment by cpatrick** [comment]
:fireworks: I'm really excited about his PR. :fireworks:

**Comment by zachmullen** [comment]
I see green check marks!

File: .travis.yml [hunk 1]
```diff
@@ -5,69 +5,57 @@ python:
...
-    - mkdir build
-    - cd build
-    - cmake ..
-    - ctest -VV -S ../cmake/travis_continuous.cmake
+  - cd $HOME
+  - rm -rf build
+  - mkdir build
+  - cd build
+  - cmake $HOME/girder
+  - ctest -V -R flow  # Only run flow test(s) for now
```
Comment (zachmullen) [inline .travis.yml]: Rather than using a test regex, we could use the new cmake variables in girder:

```
cmake -DRUN_CORE_TESTS:BOOL=OFF -DTEST_PLUGINS:STRING=flow $HOME/girder
ctest -V
```

File: plugin.cmake [hunk 2]
```diff
@@ -0,0 +1,4 @@
+add_web_client_test(
+    flow "${PROJECT_SOURCE_DIR}/plugins/flow/web_client/tests/spec/flowSpec.js"
+    ENABLEDPLUGINS "jobs romanesco flow"
```
Comment (zachmullen) [inline plugin.cmake]: We should change this from `ENABLEDPLUGINS "jobs romanesco flow"` to just `PLUGIN flow` since flow should bring the others in as dependencies.

File: .travis.yml (hunk 1 above)
Comment (danlamanna) [inline .travis.yml]: I think this will work, and I need to properly mark the Flow test as `PLUGIN flow`.

Comment (zachmullen) [inline .travis.yml]: Also we should set `-DPYTHON_COVERAGE:BOOL=OFF` so it doesn't try to do python coverage. That should give us JS coverage, though we may need to configure the threshold down with `-DJS_COVERAGE_MINIMUM_PASS:STRING=xx`.

**Comment by danlamanna** [comment]
@jeffbaumes Mind giving this a quick shot with vagrant up?

**Comment by curtislisle** [comment]
it worked first time for me! Woo hoo!   I have the tangelo screen.  Is the girder console available yet?

**Comment by jeffbaumes** [comment]
My version of ansible (2.0.1.0) seems to be hitting some errors.

First, to keep newlines in some instances, so the task in `rabbitmq-install/tasks/main.yml` was failing because it was keeping a newline so the state was `"absent\n"` instead of '"absent"`. Changing to the chomping syntax `>-` seems to fix this:

```
---

  - name: install
    apt:
        name: rabbitmq-server
        state: >-
            {{ "present" if (do_install|bool) else "absent" }}
        update_cache: true
```

But now I'm getting this:

```
TASK [girder-variables : girder | data root | default | set] *******************
task path: /Users/jeff/code/flow/devops/ansible/roles/girder-variables/tasks/main.yml:29
fatal: [flowgrant]: FAILED! => {"failed": true, "msg": "'girder_git_version' is undefined"}
```

I don't see where `girder_git_version` would be defined - is this a valid error?

**Comment by curtislisle** [comment]
My working version was on ansible 1.9.4

**Comment by danlamanna** [comment]
The Girder console should be available at /girder under the same port (8080). They're now both being served by the same CherryPy server via Girder.

**Comment by opadron** [comment]
@jeffbaumes The Ansible bits you've mentioned look like they were taken from gobig.  The `girder_git_version` variable is one defined by the `git-cache` role.

Ansible 2.0 introduced a lot of new changes, and we haven't had a chance to test the roles with it.  Among these changes, though, are significant new improvements that we'd need to reorganize the gobig roles in order to take advantage of, anyway.

So basically, try to avoid gobig + Ansible 2.0 as much as possible until we get a chance to update our roles.

**Comment by jeffbaumes** [comment]
Yes I've switched over to 1.9.4 for now and got a full vagrant up to work - @danlamanna enlightened me on the lack of support for 2.0 yet.

**Comment by jeffbaumes** [comment]
@danlamanna who came up with the clever "flowgrant" for Flow's Vagrant? I keep seeing it as "flagrant". Even though it's clever, I feel it should just be called "flow".

**Comment by danlamanna** [comment]
@jeffbaumes I came up with it in the sense that `sed s/minerva/flow/` yielded the name. I'll rename it to flow :).

**Comment by jeffbaumes** [comment]
Ah - I suppose it's even more clever for minerva with the overlapping letters :)

**Comment by curtislisle** [comment]
Great progress!  My app works nicely so far!  FYI,  .../girder works fine, but adding a slash at the end, (to make localhost:8080/girder/) doesn’t redirect to the girder console.  that was why I was confused before.

**Comment by jeffbaumes** [comment]
Yes I've found the need to get the slash right in (too) many situations.

**Comment by danlamanna** [comment]
So this is in a state where I mostly just need feedback, i.e. please try to break it and let me know if anything is inconsistent with master.

I noticed a bug when exporting to VTK formats, depending on if it seems easy or not I might just put an issue in for it since the VTK aspect of Romanesco is unused AFAIK.

**Comment by jeffbaumes** [comment]
When I do "Add Analysis" then edit the code, the font seems to be wrong and the cursor does not line up with the text. Perhaps we are overriding the CSS for the editor?

**Comment by danlamanna** [comment]
@jeffbaumes Good catch, this should be fixed.

**Comment by jeffbaumes** [comment]
There used to be a refresh icon in the upper left of the workflow steps that showed up on hover - it's currently not appearing. If you click there the functionality still works (it will go to the original analysis and reload the latest version into the workflow) but we need the button back.

screen shot 2016-03-04 at 9 41 32 am

**Comment by jeffbaumes** [comment]
In the table view, the Save button appears to be broken. It is supposed to replace the file in Girder with the updated table after editing cells, but instead I get an error

```
400 Bad Request {"message": "Attempt to load null ObjectId: ", "type": "validation"}
```

**Comment by jeffbaumes** [comment]
The scatterplot does not seem to be working. To reproduce, load `anolis.csv` and try to do a scatterplot with x and y set to SVL and attitude. I get

```
GET http://localhost:8080/scatterplot.json 404 (Not Found)
```

That JSON for the Vega library needs to be hosted statically - with it in a Girder plugin this will be somewhere other than the root I assume. Perhaps the `extras` dir?

**Comment by jeffbaumes** [comment]
I'm about done with my review - this is looking very good apart from my notes on regressions. Also, I've confirmed that this type of update works to update the Flow version:

```
git pull
vagrant provision
```

@curtislisle we'll look for your approval also before merging - please load some of your most crazy data sources and analyses and try to break it.

**Comment by jeffbaumes** [comment]
Nice! I attempted to find one, but no emojis can properly express my happiness.

**Comment by curtislisle** [comment]
The save button was  not working the pinned girder and romanesco version I installed in Chelsea's lab either.  This problem might predate the girder plug-in work.
